cmd/goplugin-identity/identity/identity.go

```diff
@@ -1,199 +1,505 @@
 package identity
 
 import (
+	"bytes"
+	"encoding/gob"
 	"fmt"
-	"time"
-
-	"github.com/boltdb/bolt"
 	"github.com/lyraproj/puppet-evaluator/eval"
+	"github.com/lyraproj/puppet-evaluator/types"
+	"github.com/lyraproj/semver/semver"
 	"github.com/lyraproj/servicesdk/grpc"
 	"github.com/lyraproj/servicesdk/service"
 	"github.com/lyraproj/servicesdk/serviceapi"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/boltdb/bolt"
 )
 
 // Identity stores identity state
 type Identity struct {
 	filename string
 }
 
+// A tuple represents an external ID with timestamp and GC status
+type tuple struct {
+	InternalID string
+	ExternalID string
+	Timestamp  time.Time
+	Era        int64
+}
+
+type storeMeta struct {
+	Version    string
+	Timestamp  time.Time
+	Era        int64
+}
+
+
+// error used internally by the identity service
+type identityError string
+
+func errorf(f string, args ...interface{}) error {
+	return identityError(fmt.Sprintf(f, args...))
+}
+
+func (ie identityError) Error() string {
+	return string(ie)
+}
+
+var metadata = []byte("metadata")
+var internalToExternal = []byte("internalToExternal")
+var externalToInternal = []byte("externalToInternal")
+var garbage = []byte("garbage")
+
+var identityStoreVersion = semver.MustParseVersion("1.0.0")
+var supportedVersions = semver.MustParseVersionRange("1.x")
+
+// Start the Identity service running
+func Start(filename string) {
+	eval.Puppet.Do(func(c eval.Context) {
+		sb := service.NewServerBuilder(c, "Default::Identity::Service")
+		id, err := NewIdentity(filename)
+		if err != nil {
+			panic(err)
+		}
+		sb.RegisterAPI(serviceapi.IdentityName, id)
+		s := sb.Server()
+		grpc.Serve(c, s)
+	})
+}
+
+// ValueTuple creates a four element Array consisting of InternalID, ExternalID, Timestamp, and GCEra.
+//
+// The Pcore type of the tuple is Tuple[String, String, Timestamp, Integer]
+func (t *tuple) ValueTuple() eval.List {
+	return types.WrapValues([]eval.Value{
```

Comment:
Why use the `eval` types here rather than a 4 element struct or array of interface{}?

Reply:
Mainly because the Identity is a remote service. At present, the wfe module calls it using invoke and as such, we either have to declare the Tuple to be a proper eval.Value type or pass it as something generic like a eval.List.